src/runtime_src/core/common/bo_cache.h

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (C) 2019 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef core_common_bo_cache_h_
+#define core_common_bo_cache_h_
+
+#include <stack>
+#include <utility>
+
+#include "xrt.h"
+#include "ert.h"
+
+namespace xrt_core {
+    // Helper typedef for std::pair. Note the elements are const so that the
+    // pair is immutable. The clients should not change the contents of cmd_bo.
+    typedef std::pair<const unsigned, struct ert_start_copybo_cmd * const> cmd_bo;
```

Comment:
this seems to be a generic mechanism for cmd bo caching, but the pointer is specific to copybo cmd?

Reply:
I indeed want to go there. However ERT cmds currently do not have the concept of base struct. Will work out the change later. Another option is to templatize the implementation with ert cmd type.

Reply:
struct ert_start_copybo_cmd is a fixed size ert cmd type. but size of other ert cmd type may not be fixed. so, it sounds like you can't use ert cmd type to templatize the implementation...
we may be able to implement a generic bo_cache to alloc and cache a bo of a certain size. it's up to user to cast the mapped bo pointer to the ert cmd they want as long as the bo size is big enough to hold the cmd data.

Reply:
The cache does not need to cache specific ert_cmd types.   In the upper level code and cmd bo is a struct that wraps the device handle and the mapped allocated BO.  Since a command BO is really just memory, a free BO can be used for any command type, which is what the upper level code does.  

Couple of general C++ comments.   
- Never use just unsigned, please specify the type.   This code will not compile on Windows
- typedefs in C++ are a thing of the past, please use 'using' which is more generic.